crates/flashblocks/p2p/src/events/dispatcher.rs

```diff
@@ -0,0 +1,171 @@
+use crate::events::NetworkPeersContext;
+
+use super::PeerEventsListener;
+use futures::{stream::FuturesUnordered, StreamExt};
+use reth_network::events::{NetworkPeersEvents, PeerEvent};
+use reth_tasks::TaskExecutor;
+use std::marker::PhantomData;
+use std::sync::Arc;
+use tokio_util::sync::CancellationToken;
+
+mod sealed {
+    /// Sealed marker trait for builder states.
+    pub trait BuilderState {}
+}
+
+use sealed::BuilderState;
+
+// Typestate markers for the builder
+// These must be public for the builder to be usable outside the crate,
+// but they are not intended for direct use (only as type parameters)
+pub struct Uninitialized;
+pub struct NetworkConfigured;
+pub struct HasListeners;
+
+impl BuilderState for Uninitialized {}
+impl BuilderState for NetworkConfigured {}
+impl BuilderState for HasListeners {}
+
+/// A typestate builder for [`PeerEventsDispatcher`].
+#[allow(private_interfaces)]
+pub struct PeerEventsDispatcherBuilder<N, State = Uninitialized>
+where
+    State: BuilderState,
+{
+    network: Option<N>,
+    listeners: Vec<Box<dyn PeerEventsListener<N>>>,
```

Comment:
I do think this may be a slightly overcomplicated solution. I was imagining basically a spawned task that periodically checks the current peer list with the list of configured trusted peers, and just emits a log for each missing peer. No listening to events required.